drivers/gpio/gpio_aw9523b.c

```diff
@@ -0,0 +1,474 @@
+
+	return err;
+}
+
+static int gpio_aw9523b_port_read_write_toggle(const struct device *dev, gpio_port_pins_t mask,
+					       gpio_port_value_t *value,
+					       enum read_write_toggle_t mode)
+{
+	const struct gpio_aw9523b_config *const config = dev->config;
+	uint8_t buf[2];
+	int err;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(aw9523b_get_lock(config->mfd_dev), K_FOREVER);
+
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 1);
+	if (!err) {
+		err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT1, buf + 1, 1);
+	}
+
+	if (err) {
+		LOG_ERR("%s: Failed to read port status (%d)", dev->name, err);
+	} else if (mode != READ) {
+		gpio_port_value_t old_value = (buf[1] << 8 | buf[0]);
+		gpio_port_value_t new_value;
+
+		if (mode == WRITE) {
+			new_value = (old_value & ~mask) | (*value & mask);
+		} else {
+			new_value = (old_value & ~mask) | (~old_value & mask);
+		}
+
+		if (new_value != old_value) {
+			buf[0] = (new_value >> 0) & 0xFF;
+			buf[1] = (new_value >> 8) & 0xFF;
+			err = i2c_burst_write_dt(&config->i2c, AW9523B_REG_OUTPUT0, buf, 2);
+			if (err) {
+				LOG_ERR("%s: Failed to set port (%d)", dev->name, err);
+			}
+		}
+	}
+
+	k_sem_give(aw9523b_get_lock(config->mfd_dev));
+
+	if (err == 0 && mode == READ) {
+		*value = (buf[1] << 8 | buf[0]);
+	}
+
+	return err;
+}
+
+static int gpio_aw9523b_port_get_raw(const struct device *dev, gpio_port_value_t *value)
+{
+	return gpio_aw9523b_port_read_write_toggle(dev, UINT16_MAX, value, READ);
+}
+
+static int gpio_aw9523b_port_set_masked_raw(const struct device *dev, gpio_port_pins_t mask,
+					    gpio_port_value_t value)
+{
+	return gpio_aw9523b_port_read_write_toggle(dev, mask, &value, WRITE);
+}
+
+static int gpio_aw9523b_port_set_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_aw9523b_port_read_write_toggle(dev, pins, &pins, WRITE);
+}
+
+static int gpio_aw9523b_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	gpio_port_value_t zero = 0;
+
+	return gpio_aw9523b_port_read_write_toggle(dev, pins, &zero, WRITE);
+}
+
+static int gpio_aw9523b_port_toggle_bits(const struct device *dev, gpio_port_pins_t pins)
+{
+	gpio_port_value_t zero = 0;
+
+	return gpio_aw9523b_port_read_write_toggle(dev, pins, &zero, TOGGLE);
+}
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+static void gpio_aw9523b_interrupt_worker(struct k_work *work)
+{
+	struct gpio_aw9523b_data *const data =
+		CONTAINER_OF(work, struct gpio_aw9523b_data, intr_worker);
+	const struct gpio_aw9523b_config *const config = data->dev->config;
+	gpio_port_value_t value, rising, falling;
+	uint8_t buf[2];
+	int err;
+
+	/* read INPUT register to deassert INTN */
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 1);
+	if (err) {
+		LOG_ERR("%s: Failed to read INPUT0 %d", data->dev->name, err);
+	}
+
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT1, buf + 1, 1);
```

Comment:
`&buf[1]`? personal preference, up to you, for some reasons I  find it more obvious when dereferencing intermediate buffer points

Also, the registeres are after one another, can't you just do `i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 2);`?

Reply:
The operation here is special; if the state of pins 0-7 changes, the interrupt will not be unlatched unless the address of INPUT0 is read, and there is the same relation between 8-15 pins and INPUT1.
At this point, we know that an interrupt has occurred, but we do not know on which side it occurred, so we need to read INPUT0 and INPUT1 independently to clear the interrupt.

Reply:
got it, thanks for explaining, consider adding a comment for future reference

Reply:
OK, I'll address it.